packages/loader/container-loader/src/deltaManager.ts

```diff
@@ -824,6 +834,24 @@ export class DeltaManager<TConnectionManager extends IConnectionManager>
             message.contents = JSON.parse(message.contents);
         }
 
+        // validate client sequence number has no gap. If there is gap, check if there were noops
+        // if there were noops, decrement the gap by number of noops and continue
+        if (this.connectionManager.lastSubmittedClientId !== undefined && this.connectionManager.lastSubmittedClientId === message.clientId) {
+            assert(this.lastClientSequenceNumber !== undefined, "lastClientSequenceNumber should not be undefined");
+            if (message.type === MessageType.NoOp){
+                this.noOpCount--;
+            }
+            const clientSeqNumGap = message.clientSequenceNumber - this.lastClientSequenceNumber - 1;
+            if (clientSeqNumGap > 0) {
+                if (this.noOpCount > 0) {
+                    this.noOpCount -= clientSeqNumGap;
+                } else {
+                    throw new Error(`gap in client sequence number :${clientSeqNumGap}`);
+                }
+            }
+        }
```

Comment:
Based on previous comment - it might be worth adding assert here that this.noOpCount >= 0. This will capture errors in logic faster.